raft.go

```diff
@@ -1749,7 +1749,7 @@ func (r *Raft) requestPreVote(rpc RPC, req *RequestPreVoteRequest) {
 	}()
 
 	// Check if we have an existing leader [who's not the candidate] and also
-	var candidate ServerAddress
+	candidate := r.trans.DecodePeer(req.GetRPCHeader().Addr)
```

Comment:
🤔 In the `requestVote` method above we do this:

```
	var candidate ServerAddress
	var candidateBytes []byte
	if len(req.RPCHeader.Addr) > 0 {
		candidate = r.trans.DecodePeer(req.RPCHeader.Addr)
		candidateBytes = req.RPCHeader.Addr
	} else {
		candidate = r.trans.DecodePeer(req.Candidate)
		candidateBytes = req.Candidate
	}
```

As far as I can tell that's the same except that we store the raw bytes for some use that we don't need here, and that for some reason we account for the case where the RPCHeader address is empty and fall back to the candidate in the message.

TBH I have so many questions about this bit of logic that I don't recall answers to if I ever did know:
 * Why do we prefer the candidate in the RPC header to the one in the message if there is always one in the message?
 * In which cases may the RPCHeader not contain an address? Maybe in-memory type transports that don't have real underlying network connections don't set that or something?

I've not had time to chase down the history on those though. If anyone else does please leave your findings here and we can understand better. But my hunch is that either the above code is redundant now and we should change both, or we should keep the same logic here as in `requestVote` (minus the unused `candidateBytes` parts).

Reply:
To add on, while implementing this change. I saw that Candidate in RequestVoteRequest is deprecated.

```go
// RequestVoteRequest is the command used by a candidate to ask a Raft peer
// for a vote in an election.
type RequestVoteRequest struct {
	RPCHeader

	// Provide the term and our id
	Term uint64

	// Deprecated: use RPCHeader.Addr instead
	Candidate []byte
```

Reply:
@k-jingyang they are deprecated but we nee do keep them for the time being to ensure retro compatibility with older versions of raft and allow a smooth upgrade from a software using an older version of raft to the new version.
I think we will need to update the fix to follow the same thing implemented in `requestVote`

```go
        var candidate ServerAddress
	if len(req.RPCHeader.Addr) > 0 {
		candidate = r.trans.DecodePeer(req.RPCHeader.Addr)
	} else {
		candidate = r.trans.DecodePeer(req.Candidate)
	}
```

Reply:
@dhiaayachi RequestPreVoteRequest does not contain any other field to indicate the addr of the peer (e.g. Candidate). I don't think we need anything other than the Addr from the RPCHeader.

```go
// RequestPreVoteRequest is the command used by a candidate to ask a Raft peer
// for a vote in an election.
type RequestPreVoteRequest struct {
	RPCHeader

	// Provide the term and our id
	Term uint64

	// Used to ensure safety
	LastLogIndex uint64
	LastLogTerm  uint64
}

```

Reply:
yes that's right! Sorry for the confusion 🤦 

@banks I think we are safe here because prevote request type only exist in the new version and an older version of raft (prior to prevote) won't process it anyway. The sender would count that as a prevote granted and carry on with the actual election.